aziot/src/init.rs

```diff
@@ -0,0 +1,867 @@
+
+	Ok((
+		keyd_config,
+		certd_config,
+		identityd_config,
+		preloaded_device_id_pk_bytes,
+	))
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display)]
+enum YesNo {
+	#[display(fmt = "Yes")]
+	Yes,
+
+	#[display(fmt = "No")]
+	No,
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display)]
+enum ProvisioningMethod {
+	#[display(fmt = "Manual provisioning with connection string (symmetric key only)")]
+	ManualConnectionString,
+
+	#[display(fmt = "Manual provisioning with symmetric key")]
+	ManualSymmetricKey,
+
+	#[display(fmt = "Manual provisioning with device identity certificate (self-signed or CA-signed)")]
+	ManualX509,
+
+	#[display(fmt = "DPS provisioning with symmetric key")]
+	DpsSymmetricKey,
+
+	#[display(fmt = "DPS provisioning with device identity certificate (self-signed or CA-signed)")]
+	DpsX509,
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display)]
+enum DeviceIdSource {
+	#[display(fmt = "Stored on the filesystem")]
+	Preloaded,
+
+	#[display(fmt = "Issued dynamically from a CA cert stored on the filesystem")]
+	LocalCa,
+
+	#[display(fmt = "Issued dynamically from an EST server")]
+	Est,
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display)]
+enum EstIdSource {
+	#[display(fmt = "Stored on the filesystem (an \"EST ID\" certificate)")]
+	Preloaded,
+
+	#[display(fmt = "Issued dynamically from the EST server using a different client certificate (a \"bootstrap EST ID\" certificate)")]
+	Bootstrap,
+}
+
+#[derive(Clone, Copy, Debug, derive_more::Display)]
+enum EstTrustedCert {
+	#[display(fmt = "My EST server's server certificate can be validated with just the CA certificates already installed on my OS")]
+	Os,
+
+	#[display(fmt = "My EST server's server certificate can only be validated with a separate CA certificate.")]
+	Separate,
+}
+
+fn choose<'a, TChoice>(
+	stdin: &mut impl std::io::BufRead,
+	question: &str,
+	choices: &'a [TChoice],
+) -> Result<&'a TChoice, crate::Error> where TChoice: std::fmt::Display {
+	println!("{}", question);
+
+	#[allow(
+		clippy::cast_possible_truncation,
+		clippy::cast_precision_loss,
+		clippy::cast_sign_loss,
+	)]
+	let max_choice_width = (choices.len() as f64).log10() as usize + 1;
+
+	println!();
+	for (i, choice) in choices.iter().enumerate() {
+		println!("{:>max_choice_width$}: {}", i + 1, choice, max_choice_width = max_choice_width);
+	}
+	println!();
+
+	let prompt_question = format!("Enter a choice [1-{}]", choices.len());
+
+	loop {
+		let answer = prompt(stdin, &prompt_question)?;
+		let answer =
+			answer.parse().ok()
+			.and_then(|choice: usize| choice.checked_sub(1))
+			.and_then(|choice| choices.get(choice));
+		if let Some(answer) = answer {
+			return Ok(answer);
+		}
+
+		print!("Invalid choice. ");
+	}
+}
+
+fn prompt(stdin: &mut impl std::io::BufRead, question: &str) -> Result<String, crate::Error> {
+	println!("{}", question);
+
+	let mut line = String::new();
+
+	loop {
+		print!("> ");
```

Comment:
A small QOL improvement would be to drop in something like https://github.com/kkawakam/rustyline
People aren't perfect, and seeing `^[[D` when you've fat-fingered a filename is never fun

Reply:
It wouldn't work, for the same reason as https://github.com/Azure/iot-identity-service/blob/1499fb72dbc2ffc89ef5a78e474db50f17672a8c/aziot/src/init.rs#L451-L453

Reply:
Oof. I guess you did that so we could mock input in tests?
That's a shame, since it'd be really nice...

Theoretically, it could be possible to use `#[cfg(test)]` to swap out the implementations, but it might be annoying.

Easy solution: make `aziot` a wrapper around a shell script that runs `rlwrap aziot-real` :wink:

Reply:
>Oof. I guess you did that so we could mock input in tests?

Yes.

Reply:
Added a trait.